rust/src/http2/http2.rs

```diff
@@ -0,0 +1,668 @@
+                    let mut tx = self.new_tx();
+                    tx.ftype = Some(head.ftype);
+                    self.transactions.push(tx);
+                    self.request_buffer.clear();
+
+                    if rlu < hl {
+                        let rl = rlu as u32;
+                        self.request_frame_size = head.length - rl;
+                        return true;
+                    } else {
+                        //toparse.len() - rlu == 9
+                        toparse = &toparse[hl - (toparse.len() - rlu)..];
+                    }
+                }
+                Err(nom::Err::Incomplete(_)) => {
+                    return true;
+                }
+                Err(_) => {
+                    self.set_event(HTTP2Event::InvalidFrameHeader);
+                    return false;
+                }
+            }
+        }
+        //then parse all we can
+        while toparse.len() > 0 {
+            match parser::http2_parse_frame_header(toparse) {
+                Ok((rem, head)) => {
+                    let mut tx = self.new_tx();
+                    tx.ftype = Some(head.ftype);
+                    self.transactions.push(tx);
+
+                    let hl = head.length as usize;
+                    if rem.len() < hl {
+                        let rl = rem.len() as u32;
+                        self.request_frame_size = head.length - rl;
+                        return true;
+                    } else {
+                        toparse = &rem[hl..];
+                    }
+                }
+                Err(nom::Err::Incomplete(_)) => {
+                    self.request_buffer.extend(toparse);
+                    return true;
+                }
+                Err(_) => {
+                    self.set_event(HTTP2Event::InvalidFrameHeader);
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    fn parse_tc(&mut self, input: &[u8]) -> bool {
+        let mut toparse = input;
+        //first consume frame bytes
+        if self.response_frame_size > 0 {
+            let ilen = input.len() as u32;
+            if self.response_frame_size >= ilen {
+                self.response_frame_size -= ilen;
+                return true;
+            } else {
+                let start = self.response_frame_size as usize;
+                toparse = &toparse[start..];
+                self.response_frame_size = 0;
+            }
+        }
+        //second extend buffer if present
+        if self.response_buffer.len() > 0 {
+            self.response_buffer.extend(toparse);
+            //parse one header locally as we borrow self and self.response_buffer
+            match parser::http2_parse_frame_header(&self.response_buffer) {
+                Ok((rem, head)) => {
+                    let hl = head.length as usize;
+                    let rlu = rem.len();
+
+                    let mut tx = self.new_tx();
```

Comment:
response should get the tx already created by request side

Reply:
Yes that is what I meant by "handle transactions with progress completion"